File: src/main/java/ru/serdtsev/homemoney/moneyoper/RecurrenceOperResource.java
package ru.serdtsev.homemoney.moneyoper;

import lombok.val;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.core.convert.ConversionService;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;
import ru.serdtsev.homemoney.account.AccountRepository;
import ru.serdtsev.homemoney.balancesheet.BalanceSheet;
import ru.serdtsev.homemoney.balancesheet.BalanceSheetRepository;
import ru.serdtsev.homemoney.common.HmException;
import ru.serdtsev.homemoney.common.HmResponse;
import ru.serdtsev.homemoney.moneyoper.model.*;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/{bsId}/recurrence-opers")
@Transactional
public class RecurrenceOperResource {
  private final MoneyOperService moneyOperService;
  private final BalanceSheetRepository balanceSheetRepo;
  private final AccountRepository accountRepo;
  private final ConversionService conversionService;

  @Autowired
  public RecurrenceOperResource(MoneyOperService moneyOperService, BalanceSheetRepository balanceSheetRepo,
       AccountRepository accountRepo, @Qualifier("conversionService") ConversionService conversionService) {
    this.moneyOperService = moneyOperService;
    this.balanceSheetRepo = balanceSheetRepo;
    this.accountRepo = accountRepo;
    this.conversionService = conversionService;
  }

  @RequestMapping
  @Transactional(readOnly = true)
  public HmResponse getList(
      @PathVariable UUID bsId,
      @RequestParam(required = false, defaultValue = "") String search) {
    try {
      BalanceSheet balanceSheet = balanceSheetRepo.findById(bsId).get();
      List<RecurrenceOperDto> list = moneyOperService.getRecurrenceOpers(balanceSheet, search)
          .sorted(Comparator.comparing(RecurrenceOper::getNextDate))
          .map(this::recurrenceOperToDto)
          .collect(Collectors.toList());
      return HmResponse.getOk(list);
    } catch (HmException e) {
      return HmResponse.getFail(e.getCode());
    }
  }

  private RecurrenceOperDto recurrenceOperToDto(RecurrenceOper recurrenceOper) {
    MoneyOper oper = recurrenceOper.getTemplate();
    String fromAccName = accountRepo.findById(oper.getFromAccId()).get().getName();
    String toAccName = accountRepo.findById(oper.getToAccId()).get().getName();
    RecurrenceOperDto dto = new RecurrenceOperDto(recurrenceOper.getId(), oper.getId(), oper.getId(),
            recurrenceOper.getNextDate(), oper.getPeriod(), oper.getFromAccId(), oper.getToAccId(), oper.getAmount(), oper.getToAmount(),
            oper.getComment(), getStringsByLabels(oper.getLabels()), oper.getCurrencyCode(), oper.getToCurrencyCode(), fromAccName,
            toAccName, oper.getType().name());
    val items = oper.getItems().stream()
            .map(item -> conversionService.convert(item, MoneyOperItemDto.class))
            .sorted(Comparator.comparing(MoneyOperItemDto::getValue))
            .collect(Collectors.toList());
    dto.setItems(items);
    return dto;
  }

  private List<String> getStringsByLabels(Collection<Label> labels) {
    return labels.stream()
        .map(Label::getName)
        .collect(Collectors.toList());
  }

  @RequestMapping("/create")
  public HmResponse create(
      @PathVariable UUID bsId,
      @RequestBody MoneyOperDto moneyOperDto) {
    BalanceSheet balanceSheet = balanceSheetRepo.findById(bsId).get();
    moneyOperService.createRecurrenceOper(balanceSheet, moneyOperDto.getId());
    return HmResponse.getOk();
  }

  @RequestMapping("/skip")
  public HmResponse skip(
      @PathVariable UUID bsId,
      @RequestBody RecurrenceOperDto oper) {
    BalanceSheet balanceSheet = balanceSheetRepo.findById(bsId).get();
    moneyOperService.skipRecurrenceOper(balanceSheet, oper.getId());
    return HmResponse.getOk();
  }

  @RequestMapping("/delete")
  public HmResponse delete(
      @PathVariable UUID bsId,
      @RequestBody RecurrenceOperDto oper) {
    try {
      BalanceSheet balanceSheet = balanceSheetRepo.findById(bsId).get();
      moneyOperService.deleteRecurrenceOper(balanceSheet, oper.getId());
      return HmResponse.getOk();
    } catch (HmException e) {
      return HmResponse.getFail(e.getCode());
    }
  }

  @RequestMapping("/update")
  public HmResponse updateRecurrenceOper(
      @PathVariable UUID bsId,
      @RequestBody RecurrenceOperDto oper) {
    try {
      BalanceSheet balanceSheet = balanceSheetRepo.findById(bsId).get();
      moneyOperService.updateRecurrenceOper(balanceSheet, oper);
      return HmResponse.getOk();
    } catch (HmException e) {
      return HmResponse.getFail(e.getCode());
    }
  }

}

File: src/main/kotlin/ru/serdtsev/homemoney/moneyoper/RecurrenceOperResource.kt
package ru.serdtsev.homemoney.moneyoper

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.core.convert.ConversionService
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import ru.serdtsev.homemoney.account.model.Reserve
import ru.serdtsev.homemoney.common.ApiRequestContextHolder
import ru.serdtsev.homemoney.common.HmException
import ru.serdtsev.homemoney.common.HmResponse
import ru.serdtsev.homemoney.moneyoper.model.*
import java.util.stream.Collectors

@RestController
@RequestMapping("/api/recurrence-opers")
@Transactional
class RecurrenceOperResource constructor(
    private val apiRequestContextHolder: ApiRequestContextHolder,
    private val moneyOperService: MoneyOperService,
    @Qualifier("conversionService") private val conversionService: ConversionService
) {
    @RequestMapping
    @Transactional(readOnly = true)
    fun getList(@RequestParam(required = false, defaultValue = "") search: String): HmResponse {
        return try {
            val list = moneyOperService.getRecurrenceOpers(apiRequestContextHolder.getBalanceSheet(), search)
                    .stream()
                    .sorted(Comparator.comparing(RecurrenceOper::nextDate))
                    .map { recurrenceOper: RecurrenceOper -> recurrenceOperToDto(recurrenceOper) }
                    .collect(Collectors.toList())
            HmResponse.getOk(list)
        } catch (e: HmException) {
            HmResponse.getFail(e.code.name)
        }
    }

    private fun recurrenceOperToDto(recurrenceOper: RecurrenceOper): RecurrenceOperDto {
        val oper = recurrenceOper.template
        val type = if (oper.type == MoneyOperType.transfer && oper.items.any { it.balance is Reserve }) {
            val operItem = oper.items.first { it.balance is Reserve }
            if (operItem.value.signum() > 0) MoneyOperType.income.name else MoneyOperType.expense.name
        } else
            oper.type.name
        val dto = RecurrenceOperDto(recurrenceOper.id, oper.id, oper.id,
                recurrenceOper.nextDate, oper.period!!, oper.comment, getStringsByTags(oper.tags), type)
        val items = oper.items
                .map { conversionService.convert(it, MoneyOperItemDto::class.java)!! }
                .sortedBy { it.value.multiply(it.sgn.toBigDecimal()) }
        dto.items = items
        return dto
    }

    private fun getStringsByTags(tags: Collection<Tag>): List<String> = tags.map(Tag::name)

    @RequestMapping("/create")
    fun create(@RequestBody moneyOperDto: MoneyOperDto): HmResponse {
        moneyOperService.createRecurrenceOper(apiRequestContextHolder.getBalanceSheet(), moneyOperDto.id)
        return HmResponse.getOk()
    }

    @RequestMapping("/skip")
    fun skip(@RequestBody oper: RecurrenceOperDto): HmResponse {
        moneyOperService.skipRecurrenceOper(apiRequestContextHolder.getBalanceSheet(), oper.id)
        return HmResponse.getOk()
    }

    @RequestMapping("/delete")
    fun delete(@RequestBody oper: RecurrenceOperDto): HmResponse {
        return try {
            moneyOperService.deleteRecurrenceOper(apiRequestContextHolder.getBalanceSheet(), oper.id)
            HmResponse.getOk()
        } catch (e: HmException) {
            HmResponse.getFail(e.code.name)
        }
    }

    @RequestMapping("/update")
    fun updateRecurrenceOper(@RequestBody oper: RecurrenceOperDto?): HmResponse {
        return try {
            moneyOperService.updateRecurrenceOper(apiRequestContextHolder.getBalanceSheet(), oper!!)
            HmResponse.getOk()
        } catch (e: HmException) {
            HmResponse.getFail(e.code.name)
        }
    }
}